### DISCUSSION TITLE
A Kotlin counterpart?

### DISCUSSION THREAD

**Discussion by elect86** [SciCatProject/scitacean#173] [Q&A]
As titled, I'd like to investigate and test the possibility to have a Kotlin counterpart interface

Since Kotlin is a statically typed safety (with excellent IDE support), we may get type-safe builders allow creating DSLs to interact with SciCat

Would you be interested in something like this?

Could I get a contact of anybody who could help/guide me?

**Comment by jl-wynen** [comment]
Interesting idea! I have personally not heard of anyone using Kotlin for scientific applications. I'd in general be interested to hear what you use it for!

Concerning support for other languages, Scitacean is designed specifically with Python in mind. So a simple port or multiple interfaces built on a common core may not be feasible. We have, however, talked about generating interfaces for multiple languages. But only in the context of a lower-level library like [pyscicat](https://github.com/SciCatProject/pyscicat). There are several SDK generators out there that can build an interface from the json schema. Would this serve your needs?

**Comment by elect86** [comment]
Mostly [Scenery](https://github.com/scenerygraphics/scenery/) and [Sciview](https://github.com/scenerygraphics/SciView/)

How does Scitacean interact with SciCat exactly? Just a plain Rest API, right? 
Because if it's so, then a port wouldn't be that hard (and long)

**Reply by jl-wynen** [reply]
> Just a plain Rest API, right?

True.

> Because if it's so, then a port wouldn't be that hard (and long)

My point was more that Scitacean does some work on top of just sending data through the API. If you are comfortable with using REST APIs, you can of course use it directly. Or, like I mentioned, auto-generate an SDK. If you do, consider pinging the people over at pyscicat to give them more motivation to do this themselves.

Scitacean is meant as a high-level tool for scientists who are usually not familiar with REST APIs or programming in general. So it makes some opinionated decisions about how the API should be used and tries to present the data in as simple a form as possible. E.g., it handles both data and metadata and (mostly) removes the need to handle data blocks. So all of that would have to be reimplemented in Kotlin.

Looking at the projects you linked, my guess is that you would mainly want to download data, right? This is straight forward for the metadata in SciCat. Concerning the data files, the solution depends on the specific facility.
If you want to upload to SciCat, it gets a little trickier if you want it to be robust because there is no built in way of doing this transactionally.

In any case, I'm happy to help out with understanding the API if you need it!

**Comment by elect86** [comment]
> Looking at the projects you linked, my guess is that you would mainly want to download data, right? 

Actually this use case is for something else, we are probing some (chemistry) scientific database for a couple of our researchers (in CASUS

> In any case, I'm happy to help out with understanding the API if you need it!

Oh, thanks, I'd love to have a quick chat in order to get a general overview and set up the project tests in order to be able to debug and port it

I'll send you an email
